Bump ginkgo from v1 to v2

Congrool opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kubeedge/kubeedge/blob/master/CONTRIBUTING.md
2. Ensure you have added or ran the appropriate tests for your PR

-->

**What type of PR is this?**

Add one of the following kinds:
/kind feature
/kind test



**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #3829 

**Special notes for your reviewer**:
This pr includes:
1. upgrades vendor, go.mod and go.sum, also replace `"github.com/onsi/ginkgo"` with `"github.com/onsi/ginkgo/v2"`.

2. deprecated `CurrentGinkgoTestDescription` with `CurrentSpecReport` as suggested in the [Migration Guide](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#changed-currentginkgotestdescription).

3. refactors the `ginkgo.Measure` spec, and implement it with `gmeasure.Experiment` as suggested in the [Migration Guide](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#removed-measure).

4. revises the ginkgo version in scripts and github actions.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

chendave reviewed (commented):
Do we have any script to build `Ginkgo`? I haven't seen any change on that in this pr.

Congrool commented:
> Do we have any script to build `Ginkgo`? I haven't seen any change on that in this pr.

@chendave Thanks, that's what i missed, and I fixed it in the forth commit which revises the ginkgo version in scripts and github actions.

chendave commented:
@Congrool thanks for the pr, I am not sure whether it's a good time to bump this for kubeedge since the change in the k8s is still pending.

If there is no such dependencies on the k8s no the Ginkgo, I think it's good to go.

chendave commented on tests/performance/loadtest/loadtest_test.go:
```diff
@@ -330,7 +330,7 @@ var _ = Describe("Application deployment test in Perfronace test EdgeNodes", fun
 		})
 
-		FMeasure("MEASURE_PERF_NODETEST_SINGLE_NODE_1: Create 100 application Deployments, Measure Pod Running time", func(b Benchmarker) {
+		FIt("MEASURE_PERF_NODETEST_SINGLE_NODE_1_WITH_1_REPLICA: Create 100 application Deployments, Measure Pod Running time", func() {
```
What's the background of this change?

Congrool replied:
Since `Measure` has been deprecated, there seems no definition nor declaration for `FMeasure` in ginkgo v2. Using `FIt` instead of `FMeasure` is just like replacing `Measure` with `It` as described in the [migration guide](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#removed-measure:~:text=replacing%20Measure%20nodes%20with%20It), though the former is not explictly writen in this guide.

chendave commented on tests/performance/loadtest/loadtest_test.go:
```diff
@@ -136,12 +137,18 @@ func PullImageInAllEdgeNodes(appDeployments []string) {
 
 //Run Test cases
-var _ = Describe("Application deployment test in Perfronace test EdgeNodes", func() {
+var _ = Describe("Application deployment test in Performance test EdgeNodes", func() {
```
Could you pls cleanup all those typos in a separate commit?

Congrool replied:
yeah, i'll do that.

Congrool commented:
@chendave thanks for your reply.
There seems no import path containing the prefix `k8s.io/test`. So I think there's no dependencies on k8s test.

Congrool commented:
Rebased this pr.

>3. refactors the ginkgo.Measure spec, and implement it with gmeasure.Experiment as suggested in the [Migration Guide](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#removed-measure).

In #3950, we removed performance tests so this migration is not needed. 

> Could you pls cleanup all those typos in a separate commit?

These typos were also removed in #3950, so do not need a separate commit.

Please take a look😃  @chendave

chendave reviewed (commented):
Just some nits, otherwise lgtm.

chendave reviewed (commented):
Just some nits, otherwise lgtm.

Congrool commented on go.sum:
```diff
@@ -861,15 +862,20 @@ github.com/onsi/ginkgo v1.12.1/go.mod h1:zj2OWP4+oCPe1qIXoGWkgMRwljMUYCdkwsT2108
 github.com/onsi/ginkgo v1.14.0/go.mod h1:iSB4RoI2tjJc9BBv4NKIKWKya62Rps+oPG/Lv9klQyY=
 github.com/onsi/ginkgo v1.16.4 h1:29JGrr5oVBm5ulCWet69zQkzWipVXIol6ygQUe/EzNc=
 github.com/onsi/ginkgo v1.16.4/go.mod h1:dX+/inL/fNMqNlz0e9LfyB9TswhZpCVdJM/Z6Vvnwo0=
+github.com/onsi/ginkgo/v2 v2.1.3/go.mod h1:vw5CSIxN1JObi/U8gcbwft7ZxR2dgaR70JSE3/PpL4c=
```
I've runned `go mod tidy`. It seems that no entries can be removed.

Congrool commented on tests/e2e/scripts/execute.sh:
```diff
@@ -22,15 +22,30 @@ cd $workdir
-    sudo cp $GOPATH/bin/ginkgo /usr/local/bin/
-)
+check_ginkgo() {
+    # check if installed
```
Yes, I check the version in the next step. 

It follows steps like:
1. check if ginkgo cmd exists, otherwise install ginkgo v2
2. check the ginkgo version of existing ginkgo cmd, if it's not ginkgo v2, reinstall it.

chendave replied:
I just meant that your comment there could be more accurate. :)

Congrool commented on tests/e2e/scripts/execute.sh:
```diff
@@ -22,15 +22,30 @@ cd $workdir
+    # Assuming ginkgo version output format is: 
+    # Ginkgo Version 1.14.0
+    local majorVer=($(echo ${ginkgo_version_output[2]} | sed "s/\./ /g")[0])
+    if [[ ${majorVer} -ne 2 ]]; then
```
good idea

Congrool commented on tests/e2e/scripts/execute.sh:
```diff
@@ -22,15 +22,30 @@ cd $workdir
-    go install github.com/onsi/ginkgo/ginkgo@latest
-    sudo cp $GOPATH/bin/ginkgo /usr/local/bin/
-)
+check_ginkgo() {
```
fixed

Congrool commented:
@chendave Thanks for your suggestions, and all replied. Please take a look :).

chendave commented on go.sum:
```diff
@@ -861,15 +862,20 @@ github.com/onsi/ginkgo v1.12.1/go.mod h1:zj2OWP4+oCPe1qIXoGWkgMRwljMUYCdkwsT2108
 github.com/onsi/ginkgo v1.14.0/go.mod h1:iSB4RoI2tjJc9BBv4NKIKWKya62Rps+oPG/Lv9klQyY=
 github.com/onsi/ginkgo v1.16.4 h1:29JGrr5oVBm5ulCWet69zQkzWipVXIol6ygQUe/EzNc=
 github.com/onsi/ginkgo v1.16.4/go.mod h1:dX+/inL/fNMqNlz0e9LfyB9TswhZpCVdJM/Z6Vvnwo0=
```
Just like what we did in the https://github.com/kubernetes/kubernetes/pull/109111, we need to figure out why ginkgo v1 is still here, and try to clean all them up.

gy95 replied:
LGTM :)
do we need to fix this comment?

Congrool replied:
Yes, we discussed in comments bellow. And the conclusion is that, currently we cannot prune ginkgo v1 here. @gy95

Congrool commented:
The ginkgo entries we have in go.sum:
```plain
github.com/onsi/ginkgo v0.0.0-20151202141238-7f8ab55aaf3b/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
github.com/onsi/ginkgo v0.0.0-20170829012221-11459a886d9c/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
github.com/onsi/ginkgo v1.6.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
github.com/onsi/ginkgo v1.10.1/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
github.com/onsi/ginkgo v1.10.3/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
github.com/onsi/ginkgo v1.11.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
github.com/onsi/ginkgo v1.12.1/go.mod h1:zj2OWP4+oCPe1qIXoGWkgMRwljMUYCdkwsT2108oapk=
github.com/onsi/ginkgo v1.14.0/go.mod h1:iSB4RoI2tjJc9BBv4NKIKWKya62Rps+oPG/Lv9klQyY=
github.com/onsi/ginkgo v1.16.4 h1:29JGrr5oVBm5ulCWet69zQkzWipVXIol6ygQUe/EzNc=
github.com/onsi/ginkgo v1.16.4/go.mod h1:dX+/inL/fNMqNlz0e9LfyB9TswhZpCVdJM/Z6Vvnwo0=
github.com/onsi/ginkgo/v2 v2.1.3/go.mod h1:vw5CSIxN1JObi/U8gcbwft7ZxR2dgaR70JSE3/PpL4c=
github.com/onsi/ginkgo/v2 v2.1.4 h1:GNapqRSid3zijZ9H77KrgVG4/8KqiyRsxcSxe+7ApXY=
github.com/onsi/ginkgo/v2 v2.1.4/go.mod h1:um6tUpWM/cxCK3/FK8BXqEiUMUwRgSM4JXG47RKZmLU=
```

I used `go mod graph` to get who uses these different versions of ginkgo.

The dependencies for each are:

Packages depend on github.com/onsi/ginkgo@v0.0.0-20151202141238-7f8ab55aaf3b:
```plain
github.com/containernetworking/plugins@v0.8.6 
```

Packages depend on github.com/onsi/ginkgo@v0.0.0-20170829012221-11459a886d9c: 
```plain
k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65
k8s.io/kube-openapi@v0.0.0-20211109043538-20434351676c
k8s.io/kube-openapi@v0.0.0-20201113171705-d219536bb9fd
k8s.io/kube-openapi@v0.0.0-20210421082810-95288971da7e
```

Packages depend on github.com/onsi/ginkgo@v1.6.0:
```plain
github.com/onsi/gomega@v1.8.1
github.com/onsi/gomega@v1.7.1
github.com/onsi/gomega@v1.7.0
```

Packages depend on github.com/onsi/ginkgo@v1.10.0:
```plain
github.com/containerd/continuity@v0.0.0-20200710164510-efbc4488d8fe 
```

Packages depend on github.com/onsi/ginkgo@v1.10.3:
```plain
github.com/containerd/go-cni@v1.0.2
github.com/containerd/go-cni@v1.0.1
```

Packages depend on github.com/onsi/ginkgo@v1.11.0:
```plain
github.com/kubeedge/viaduct@v0.0.0
k8s.io/apimachinery@v0.20.6
k8s.io/apimachinery@v0.20.10
k8s.io/apimachinery@v0.20.1
k8s.io/apimachinery@v0.20.4
```

Packages depend on github.com/onsi/ginkgo@v1.12.1:
```plain
github.com/containernetworking/plugins@v0.9.1
github.com/onsi/gomega@v1.10.1
github.com/onsi/ginkgo@v1.12.1
github.com/onsi/ginkgo@v1.12.1
github.com/onsi/ginkgo@v1.12.1
github.com/onsi/gomega@v1.10.3
```
... 

I have no idea of how to remove these dependencies on different versions of ginkgo. And its seems that we should not care about it. Golang will recursively add direct and indirect dependencies into the go.sum, see the [blog](https://golangbyexample.com/go-mod-sum-module/#:~:text=This%20file%20lists%20down%20the%20checksum%20of%20direct%20and%20indirect%20dependency%20required%20along%20with%20the%20version.).

For example:
kubeedge directly import [helm.sh/helm/v3@3.7.2](https://github.com/kubeedge/kubeedge/blob/master/go.mod#L44), and the import link is 
[helm.sh/helm/v3@3.7.2](https://github.com/kubeedge/kubeedge/blob/master/go.mod#L44) -> [github.com/containerd/containerd@v1.5.7](https://github.com/helm/helm/blob/v3.7.2/go.mod#L13) -> [github.com/containerd/go-cni@v1.0.2](https://github.com/containerd/containerd/blob/v1.5.7/go.mod#L14) -> [github.com/onsi/ginkgo@v1.10.3](https://github.com/containerd/go-cni/blob/v1.0.2/go.mod#L6)

Thus we have `github.com/onsi/ginkgo v1.10.3/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=` in go.sum of KubeEdge. Based on this machanism, this entry is not introduced by kubeedge itself, so we should not care about it.

But I have to say, I'm not familiar with the how to generate go.sum. Maybe there're some misunderstandings. What do you think? @chendave

chendave commented:
A similar issue was found here: https://github.com/kubernetes/kube-openapi/pull/306, and bump golang to 1.18 solve that problem.

Maybe we can have a try?

chendave commented:
As a reference: https://github.com/kubernetes/kube-openapi/pull/302/files

chendave commented:
@Congrool  FYI, https://github.com/kubernetes/kubernetes/pull/109111 was merged.

Do you have any update so far?

Congrool commented:
Have not yet, I'll do with it this weekend.

Congrool commented:
I tried to [update go1.18](https://github.com/kubeedge/kubeedge/pull/3835/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6) to prune the ginkgo v1 dependency, but it seems to have no effect. I guess it is because they are introduced by 3rd party package instead of ginkgo/gomega themselves.

I wrote a [tool](https://github.com/Congrool/kubeedge/commit/c0d299c74ee5182f05214e9cda419955a8911e23#diff-11f45327a2a30fc05f67818367067456675207478a422661635c41a9dc754e31) to generate invoke chains of each ginkgo version according to the ouput of `go mod graph`, and the result can be found at [here](https://github.com/Congrool/kubeedge/commit/c0d299c74ee5182f05214e9cda419955a8911e23).
Each entry has the format:
```plaintext
github.com/onsi/ginkgo@xxxx->...->github.com/kubeedge/kubeedge
```
`->` means "invoked by".

I found that most dependency of ginkgo v1 is introduced by 3rd party package, such as `helm.sh/helm/v3@v3.7.2`.

I also verfy it on the kube-openapi repo at `commit: 1062c7a`(it is a commit before prs of update go1.18 and bump ginkgo v2), running the [tool](https://github.com/Congrool/kubeedge/commit/c0d299c74ee5182f05214e9cda419955a8911e23#diff-11f45327a2a30fc05f67818367067456675207478a422661635c41a9dc754e31) to figure out why we can prune ginkgo v1 dependency through update go1.18. First, I just `go get github.com/onsi/ginkgo/v2@v2.1.4` with `go1.16` to reproduce the case what disscussed in pr kubernetes/kube-openapi#302, and then run the tool. The invoke chains of ginkgo in kube-openapi is like:
```plaintext
#ginkgo@1.6.0
github.com/onsi/ginkgo@v1.6.0->github.com/onsi/gomega@v1.7.1->github.com/onsi/ginkgo@v1.12.1->github.com/onsi/gomega@v1.10.1->github.com/onsi/ginkgo@v1.16.4->github.com/onsi/gomega@v1.17.0->github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->k8s.io/kube-openapi
github.com/onsi/ginkgo@v1.6.0->github.com/onsi/gomega@v1.7.1->github.com/onsi/ginkgo@v1.12.1->github.com/onsi/gomega@v1.10.1->github.com/onsi/ginkgo@v1.16.4->github.com/onsi/gomega@v1.17.0->github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->github.com/onsi/ginkgo/v2@v2.1.4->k8s.io/kube-openapi

#ginkgo@1.21.1
github.com/onsi/ginkgo@v1.12.1->github.com/onsi/gomega@v1.10.1->github.com/onsi/ginkgo@v1.16.4->github.com/onsi/gomega@v1.17.0->github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->k8s.io/kube-openapi
github.com/onsi/ginkgo@v1.12.1->github.com/onsi/gomega@v1.10.1->github.com/onsi/ginkgo@v1.16.4->github.com/onsi/gomega@v1.17.0->github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->github.com/onsi/ginkgo/v2@v2.1.4->k8s.io/kube-openapi

#ginkgo@1.16.4
github.com/onsi/ginkgo@v1.16.4->github.com/onsi/gomega@v1.17.0->github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->k8s.io/kube-openapi
github.com/onsi/ginkgo@v1.16.4->github.com/onsi/gomega@v1.17.0->github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->github.com/onsi/ginkgo/v2@v2.1.4->k8s.io/kube-openapi

#ginkgo@v2.1.3
github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->k8s.io/kube-openapi
github.com/onsi/ginkgo/v2@v2.1.3->github.com/onsi/gomega@v1.19.0->github.com/onsi/ginkgo/v2@v2.1.4->k8s.io/kube-openapi

#ginkgo@2.1.4
github.com/onsi/ginkgo/v2@v2.1.4->k8s.io/kube-openapi
```

We can see that all these dependencies are between ginkgo and gomega, without 3rd party. So, we can prune them through the transitive dependency pruning feature of go1.18, I guess, through I've not found any doc to demonstrate it.

As a conclusion, what we do in kubernetes/kube-openapi#302 may not work in KubeEdge.

chendave commented:
> Packages depend on github.com/onsi/ginkgo@v0.0.0-20170829012221-11459a886d9c:

> k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65
> k8s.io/kube-openapi@v0.0.0-20211109043538-20434351676c
> k8s.io/kube-openapi@v0.0.0-20201113171705-d219536bb9fd
> k8s.io/kube-openapi@v0.0.0-20210421082810-95288971da7e


Since we are basing on the latest release of those deps, it's not easy for us to get rid of ginkgo v1 by far, which mean both v1 and v2 will be listed as the dep for kubeedge.

But wait for all those project to bump to the Ginkgo will not be that fast.

One thing I am not sure is that even we still need Ginkgo v1, but it's not listed in `go.mod`.

e.g. 
```
# go mod why github.com/onsi/ginkgo
# github.com/onsi/ginkgo
github.com/kubeedge/kubeedge/cloud/cmd/controllermanager/app
sigs.k8s.io/controller-runtime/pkg/client/config
sigs.k8s.io/controller-runtime/pkg/client/config.test
github.com/onsi/ginkgo
```


Besides, 

`go mod tidy` would remove the below packages,
```
        deleted:    vendor/github.com/kubeedge/beehive
        deleted:    vendor/github.com/kubeedge/viaduct
```


Anyway, this pr helps us to remove the Ginkgo v1 from the folder of vendor.
@fisherxu  WDYT?

chendave commented:
Can we remove this dep "sigs.k8s.io/controller-runtime" from the project?

chendave commented:
My suggestion is,

- bump "kube-openapi" to the latest since that project has been upgraded to v2.
- need some investigation on the "sigs.k8s.io/controller-runtime", if we can remove this dep or get some info on when that project could be bumped to the v2.

Congrool commented:
> Can we remove this dep "sigs.k8s.io/controller-runtime" from the project?

I can reply for this. The new component, called ControllerManager is developed based on sigs.k8s.io/controller-runtime, introduced in pr #3719.  It's directly needed, so i think it cannot be removed. 

We use "sigs.k8s.io/controller-runtime@v0.10.3" not the lates v0.12.3 because of the kubernetes dependency. `v0.10.3` is the latest version using the kubernetes v1.22 which KubeEdge is currently dependent on. So, we cannot bump `controller-runtime` unless KubeEdge can support higher version for Kubernetes.

chendave commented:
How about the openapi, here is the reference: https://github.com/kubernetes/kubernetes/pull/110831

Congrool commented:
Would we still need to update go1.18 in this pr? It seems not help us to prune the ginkgo dependencies. Maybe we can update it in another pr if we need.

chendave commented:
> Maybe we can update it in another pr if we need.

+1

gy95 commented on tests/e2e/scripts/execute.sh:
```diff
@@ -22,15 +22,32 @@ cd $workdir
+    which ginkgo &> /dev/null || (
+        echo "ginkgo is not found, install first"
+        go install -mod=mod github.com/onsi/ginkgo/v2/ginkgo
+        sudo cp $GOPATH/bin/ginkgo /usr/local/bin/
```
I mean we can return here directly, no need to continue to line34-line43 to install it again

Congrool replied:
fixed

Congrool commented:
An analysis of kube-openapi dependencies here.

First of all, a report of which pkg that KubeEdge directly import causes the dependencies on these versions of kube-openapi:

```plaintext
# pkgs depends on k8s.io/kube-openapi@v0.0.0-20201113171705-d219536bb9fd
sigs.k8s.io/apiserver-network-proxy@v0.0.27
github.com/containerd/containerd@v1.5.10
helm.sh/helm/v3@v3.7.2

# pkgs depends on k8s.io/kube-openapi@v0.0.0-20210421082810-95288971da7e
k8s.io/cli-runtime@v0.22.6
helm.sh/helm/v3@v3.7.2
k8s.io/kubernetes@v1.22.6
sigs.k8s.io/controller-runtime@v0.10.3

#pkgs depends on k8s.io/kube-openapi@v0.0.0-20211109043538-20434351676c
k8s.io/code-generator@v0.22.6
k8s.io/kube-scheduler@v0.22.6
k8s.io/apiserver@v0.22.6
k8s.io/cli-runtime@v0.22.6
k8s.io/kubelet@v0.22.6
k8s.io/apiextensions-apiserver@v0.22.6
helm.sh/helm/v3@v3.7.2
k8s.io/cluster-bootstrap@v0.22.6
k8s.io/api@v0.22.6
k8s.io/cloud-provider@v0.22.6
k8s.io/kubernetes@v1.22.6
k8s.io/client-go@v0.22.6
k8s.io/component-base@v0.22.6
k8s.io/csi-translation-lib@v0.22.6
k8s.io/apimachinery@v0.22.6

# pkg depends on k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65
Only KubeEdge itself
```

As we can see above, KubeEdge mainly depends on kubernetes@v0.22.6. We cannot remove the ginkgo dependency `github.com/onsi/ginkgo@v0.0.0-20170829012221-11459a886d9c` if we do not bump the kubernetes dependency of KubeEdge, because we will always depend on it if we depend on something like `k8s.io/apiserver@v0.22.6`.

So, if we just bump kube-openapi to its latest version in go.mod. Only this will be changed
```patch
- # pkg depends on k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65
+ # pkg depends on k8s.io/kube-openapi@latest
Only KubeEdge itself
```
Thus, in summary, bump kube-openapi does not help us to remove the ginkgo dependencies. @chendave 

BTW, it seems that we can change `k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65` to `k8s.io/kube-openapi@v0.0.0-20211109043538-20434351676c` in go.mod to reduce some entries of kube-openapi in go.sum 😄 .

chendave commented:
> As we can see above, KubeEdge mainly depends on kubernetes@v0.22.6. 

ACK, the problem could be mitigated after bumping to the kubernetes 1.25, but for now, could you please build the binary and check the size of each binaries, I'd hope this will not expand the size of binary too much.

Congrool commented:
Binary sizes before this pr
```plaintext
$ ls -sh _output/local/bin/
total 389M
28M admission  37M controllermanager  64M edgecore  11M edgesite-agent   32M iptablesmanager
42M cloudcore  17M csidriver          64M edgemark  34M edgesite-server  65M keadm
```

Binary sizes after this pr
```plaintext
$ ls -sh _output/local/bin/
total 389M
28M admission  37M controllermanager  64M edgecore  11M edgesite-agent   32M iptablesmanager
42M cloudcore  17M csidriver          64M edgemark  34M edgesite-server  65M keadm
``` 

Glad to see that there'no obvious change on binary size.

chendave commented:
/approve

pls create an issue to track this, we need to cleanup the v1 dep once k8s is upgrade to 1.25.

chendave commented:
leaving lgtm to @gy95 or @fisherxu for the double check.

chendave commented:
/hold

gy95 commented:
/lgtm

Congrool commented:
The issue #4049 has be created to trace the left problem of this pr.

chendave commented:
thanks @Congrool !

/hold cancel
/lgtm
